ui/src/shared/components/HistogramChartBars.tsx

```diff
@@ -1,27 +1,175 @@
-import React, {PureComponent} from 'react'
+import React, {PureComponent, MouseEvent} from 'react'
 import _ from 'lodash'
 import {ScaleLinear, ScaleTime} from 'd3-scale'
+import {color} from 'd3-color'
 
-import {HistogramData, HistogramDatum} from 'src/types/histogram'
+import {getDeep} from 'src/utils/wrappers'
+
+import {
+  HistogramData,
+  HistogramDatum,
+  HoverData,
+  TooltipAnchor,
+  ColorScale,
+} from 'src/types/histogram'
 
 const BAR_BORDER_RADIUS = 4
 const BAR_PADDING_SIDES = 4
+const HOVER_BRIGTHEN_FACTOR = 0.4
+const TOOLTIP_HORIZONTAL_MARGIN = 5
+const TOOLTIP_REFLECT_DIST = 100
+
+const getBarWidth = ({data, xScale, width}): number => {
+  const dataInView = data.filter(
+    d => xScale(d.time) >= 0 && xScale(d.time) <= width
+  )
+  const barCount = Object.values(_.groupBy(dataInView, 'time')).length
+
+  return Math.round(width / barCount - BAR_PADDING_SIDES)
+}
+
+type SortFn = (a: HistogramDatum, b: HistogramDatum) => number
+
+const getSortFn = (data: HistogramData): SortFn => {
+  const counts = {}
+
+  for (const d of data) {
+    if (counts[d.group]) {
+      counts[d.group] += d.value
+    } else {
+      counts[d.group] = d.value
+    }
+  }
+
+  return (a, b) => counts[b.group] - counts[a.group]
+}
+
+interface BarGroup {
+  key: string
+  clip: {
+    x: number
+    y: number
+    width: number
+    height: number
+  }
+  bars: Array<{
+    key: string
+    group: string
+    x: number
+    y: number
+    width: number
+    height: number
+    fill: string
+  }>
+  data: HistogramData
+}
+
+const getBarGroups = ({
+  data,
+  width,
+  xScale,
+  yScale,
+  colorScale,
+  hoverData,
+}): BarGroup[] => {
+  const barWidth = getBarWidth({data, xScale, width})
+  const sortFn = getSortFn(data)
+  const visibleData = data.filter(d => d.value !== 0)
+  const timeGroups = Object.values(_.groupBy(visibleData, 'time'))
+
+  for (const timeGroup of timeGroups) {
+    timeGroup.sort(sortFn)
+  }
+
+  let hoverDataKeys = []
+
+  if (!!hoverData) {
+    hoverDataKeys = hoverData.data.map(h => h.key)
+  }
+
+  return timeGroups.map(timeGroup => {
+    const time = timeGroup[0].time
+    const x = xScale(time) - barWidth / 2
+    const total = _.sumBy(timeGroup, 'value')
+
+    const barGroup = {
+      key: `${time}-${total}-${x}`,
+      clip: {
+        x,
+        y: yScale(total),
+        width: barWidth,
+        height: yScale(0) - yScale(total) + BAR_BORDER_RADIUS,
+      },
+      bars: [],
+      data: timeGroup,
+    }
+
+    let offset = 0
+
+    timeGroup.forEach((d: HistogramDatum) => {
```

Comment:
We could lift this above the definition of `barGroup` and defining `bars` as 

```
const bars = timeGroup.map...
```

and then just return the `barGroup` object with ` bars`

Reply:
although the `forEach` does draw my attention to the `offset`